### ISSUE TITLE
RatBagd resets settings on computer restart

### ISSUE THREAD

**Issue by codedcosmos** [libratbag/libratbag#596]
This issue was originally opened here:
https://github.com/libratbag/piper/issues/282

I seem to only be able to retain my settings for the current session. I can hit apply, and it seems to save alright. It will always say I have unsaved settings, but when I reopen the window it seems to have saved fine.

I can use those settings for the entire session. But when I restart, the settings revert to original.

I am using the MX master 2s, and am also trying to use the unifying receiver, which doesn't seem to work consistently with the mouse and piper.

Additionally, is there any way to apply settings with ratbagd and not piper?

**Comment by bentiss** [comment]
That's a limitation from the mouse and the choices we made with libratbag:
- the mouse doesn't have a persistent state like the gaming mice we are dealing with
- libratbag doesn't have a persistent state either as we rely on the mouse to store its data (to avoid race conditions with Windows or anyone else changing the settings in our back).

So there is no easy way to solve this besides having a custom script run after the mouse is plugged in (like a udev rule).

> Additionally, is there any way to apply settings with ratbagd and not piper?

`ratbagctl` should be your friend there. Run `ratabgctl list`, then `ratbagctl info eventX` and `ratbagctl --help` to have an idea on how to change the values in a script.

**Comment by andigandi** [comment]
There is an other problem with the MX Master 2S. It also loses its settings, when it enters standby mode, which happens after a couple of minutes not using it. Is there a way of registering a mouse wakeup with udev? And do you have an idea, how Logitechs Windows software manages the settings on such devices?

**Comment by whot** [comment]
My guess is the windows software detects when the mouse wakes up, maybe there's some event that comes from it. In response, it can apply the same settings again. Not impossible to implement in ratbagd, it would just need someone with sufficient motivation to get started.

**Comment by devxpy** [comment]
@whot Does this need to be something implemented inside libratbagd, or can it be done inside 3rd party tools written in easier languages like python?

**Comment by aleho** [comment]
@whot Yes, there's actually an event. According to https://github.com/pwr/Solaar/issues/467#issuecomment-432196257 the receiver sends it as soon as the mouse wakes up. I'm currently using an udev rule to set DPI via `solaar config`, but that's more of a hackish workaround and also only good for single-user systems.

**Comment by aleho** [comment]
@whot Yes, there's actually an event. According to https://github.com/pwr/Solaar/issues/467#issuecomment-432196257 the receiver sends it as soon as the mouse wakes up. I'm currently using an udev rule to set DPI via `solaar config`, but that's more of a hackish workaround and also only good for single-user systems.

**Comment by devxpy** [comment]
@aleho  What about Bluetooth?

**Comment by bentiss** [comment]
>  @aleho What about Bluetooth?

The Logitech HID++ protocol is the same whether you use Bluetooth, USB or wireless with the unifying receiver. And yes, we are aware of the `connected` event, and it's also publicly documented (page 9, section 3.2 of https://lekensteyn.nl/files/logitech/logitech_hidpp10_specification_for_Unifying_Receivers.pdf).

The initial purpose of libratbag, ratbagd and piper was to configure gaming mice that have an onboard memory. It turns out that the Logitech mice can also be using libratbag, but ratbagd was not designed to listen to the events of the mouse. It just opens the communication channel, configure the mouse and closes it. There is no long standing reading of the incoming events, so ratbagd is not ready today to handle such connection/disconnection events.

And as @whot said, we need someone to work on this, and I wouldn't be against a PR that implement such a feature (but it will introduce a debate, that is for sure). And yes, IMO, ratbagd should be where you do that, because it can listen to the hidraw nodes, while a user script needs root for that.

Note also that this would suddenly turns ratbagd into a keylogger, as it will listen to the events coming in from the mouse/keyboards, and that's where there is a debate. (Or any script you can think of that reads the HID++ events would be a keylogger too).

**Comment by andigandi** [comment]
> I'm currently using an udev rule to set DPI via `solaar config`

Could you share your udev rule? I let `udevadm monitor` run for some time but could not catch anything useful. The only thing I could record were some `/devices/pci0000:00/0000:00:14.0/usb3/3-9/3-9.4/3-9.4.1/3-9.4.1.3/3-9.4.1.3:1.2/0003:046D:C52B.0009/0003:046D:4069.000A/power_supply/hidpp_battery_1` events but I am not quite sure when those were issued. You could easily replace `solaar` with `ratbagctl` commands.

**Comment by devxpy** [comment]
@bentiss I suppose [Solaar](https://github.com/pwr/Solaar) could be a better place to take this discussion. 

They already support dpi setting and smart shift sensitivity as a permanent setting. If you say that the HID interface is same, maybe adding bluetooth to solaar wouldn't be a huge pain.

Thanks for all the work on this BTW!

**Comment by nrayamajhee** [comment]
This issue also affects MX Anywhere 2S. <https://github.com/libratbag/piper/issues/330>
@aleho Can you please share the udev rule?
I tried to write my own. I have no prior experience with udev rules. I wrote this looking at the guide on on arch linux's wiki,  but I couldn't get it to work.
```bash
ACTION=="add"\
, SUBSYSTEM=="power_supply"\
, ENV{POWER_SUPPLY_MODEL_NAME}=="Wireless Mobile Mouse MX Anywhere 2S"\
, ENV{POWER_SUPPLY_ONLINE}=="1"\
, ENV{POWER_SUPPLY_STATU}=="Discharging"\
, RUN+="ratbagctl Logitech\ MX\ Anywhere\ 2S dpi set 3800"
```
Even the hackish workaround would help.

**Comment by FFY00** [comment]
@nrayamajhee see https://wiki.archlinux.org/index.php/udev#Accessing_firmware_programmers_and_USB_virtual_comm_devices, you probably need to change the `SUBSYSTEMS` entry to `hidraw`.

**Comment by nrayamajhee** [comment]
It seems like my udev rule is actually getting triggered. I tried with hidraw and nothing happened.
What I noticed was that systemd-udev does not run the command. I get the following message in log:
`systemd-udevd[857]: Process 'ratbagctl Logitech\ MX\ Anywhere\ 2S dpi set 3800' failed with exit code 1`
I suspected it was because of the command, but apparently it's not. No matter what command I put in `RUN+=` I get the same error saying the process failed with exit code 1.

**Comment by nrayamajhee** [comment]
So I finally fixed it: the following udev rule worked:
```bash
ACTION=="change"\
, SUBSYSTEM=="power_supply"\
, ENV{POWER_SUPPLY_MODEL_NAME}=="Wireless Mobile Mouse MX Anywhere 2S"\
, ENV{POWER_SUPPLY_ONLINE}=="1"\
, ENV{POWER_SUPPLY_STATUS}=="Discharging"\
, RUN+="/usr/bin/su root -c 'ratbagctl Logitech\ MX\ Anywhere\ 2S dpi set 2400'"
```
Thanks for helping out.

**Comment by dlbnco** [comment]
This also happens with the MX Master 3.

**Comment by manojmenon** [comment]
This issue is also there in MX Master 3. - I noticed it happens in Windows 10 however, the mouse works well for  Mac (OS 10.15.1)

**Comment by Roshi88** [comment]
Hi, i'm facing the same issue with bluetooth mouse Logitech M590, i think the reason is the same, but i'm having issues in the keys programmability, i use piper to swap the "backward" and "forward" keys.

Do you know which command shall i give, as a workaround, instead of "2S dpi set XXXX" or where i can search the command list?

Thanks in advance

**Comment by nrayamajhee** [comment]
@Roshi88 I think `ratbagctl --help` should show you some info on what commands are available.
I think its something like: `ratbagctl DEVICE button ONE action set button ANOTHER`. 
But your device name is going to be different so your udev rule needs a different device identifier. Plus, you mention using bluetooth so I can't say for sure if that udev rule is going to work.

**Comment by krmbzds** [comment]
I use the following udev rule for to set DPI on my `Logitech MX Anywhere 2S` 

```
ACTION=="change", SUBSYSTEM=="power_supply", ATTR{online}=="1", ATTRS{idVendor}=="046d", ATTRS{idProduct}=="c52b", RUN+="/usr/bin/ratbagctl 'Logitech MX Anywhere 2S' dpi set 400"
```

The udev rule triggers fine (captures power on, wake up events)  but `ratbagd` fails on subsequent runs unless I run `sudo killall ratbagd`, which makes the whole thing kind of useless.

**Comment by chicolucio** [comment]
Same issue with Logitech MX Master 3S.

**Comment by XorvatoAmigos** [comment]
Maybe this will be useful to someone.

Logitech g305 Lightspeed mouse, configured it with Piper, changed the polling rate and DPI, the settings were not saved, after rebooting or disconnecting the mouse everything went back to the original state. 

Got curious, booted into Windows, decided to use the Onboard Memory Manager utility (Logitech) to change the settings, after disconnecting the mouse or rebooting the same thing happened, which made me wonder even more. 

Booted up GHub or Logitech Gaming Software, found the profile set to low performance mode (frequencies), enabled high performance mode, saved.

Booted into Linux, checked changing and saving settings with Piper, everything works as it should.
